libsolidity/analysis/TypeChecker.cpp

```diff
@@ -426,6 +429,110 @@ bool TypeChecker::visit(FunctionDefinition const& _function)
 		else if (_function.libraryFunction())
 			m_errorReporter.typeError(7801_error, _function.location(), "Library functions cannot be \"virtual\".");
 	}
+	if (_function.usableAsSuffix())
+	{
+		if (_function.stateMutability() != StateMutability::Pure)
+			m_errorReporter.typeError(
+				1716_error,
+				_function.location(),
+				"Only pure functions can be used as literal suffixes"
+			);
+
+		switch (_function.parameterList().parameters().size())
+		{
+		case 1:
+		{
+			auto&& parameterType = _function.parameterList().parameters()[0]->type();
+
+			if (dynamic_cast<FixedPointType const*>(parameterType))
+				m_errorReporter.typeError(
+					2699_error,
+					_function.parameterList().parameters()[0]->location(),
+					"Parameters of fixed-point types are not allowed in literal suffix functions. "
+					"To support fractional literals the suffix function must accept two integer arguments "
+					"(mantissa and exponent) that such literals can be decomposed into."
+				);
+
+			if (
+				!TypeProvider::boolean()->isImplicitlyConvertibleTo(*parameterType) &&
+				// ASSUMPTION: There are no address payable literals.
+				!TypeProvider::address()->isImplicitlyConvertibleTo(*parameterType) &&
+				// ASSUMPTION: Literal 1 is implicitly convertible to any integer type.
+				!TypeProvider::rationalNumber(1)->isImplicitlyConvertibleTo(*parameterType) &&
+				// ASSUMPTION: bytes1 is implicitly convertible to any fixed-bytes type.
+				!TypeProvider::fixedBytes(1)->isImplicitlyConvertibleTo(*parameterType) &&
+				!TypeProvider::stringLiteral("a")->isImplicitlyConvertibleTo(*parameterType)
+			)
+				m_errorReporter.typeError(
+					2998_error,
+					_function.parameterList().parameters()[0]->location(),
+					"This literal suffix function is not usable as a suffix because no literal is "
+					"implicitly convertible to its parameter type."
+				);
+
+			break;
+		}
+
+		case 2:
+		{
+			auto&& mantissaType = dynamic_cast<IntegerType const*>(_function.parameterList().parameters()[0]->type());
+			auto&& exponentType = dynamic_cast<IntegerType const*>(_function.parameterList().parameters()[1]->type());
```

Comment:
```suggestion
			auto const* mantissaType = dynamic_cast<IntegerType const*>(_function.parameterList().parameters()[0]->type());
			auto const* exponentType = dynamic_cast<IntegerType const*>(_function.parameterList().parameters()[1]->type());
```
Optional nitpick - but yeah, while I'm a fan of ``auto`` in general and of ``auto&&`` in particular, I actually find it easier on the eye here to make it clear it's a ``const*`` :-).

Reply:
Haha, for me this is actually one of the few cases where omitting the `const*` is an improvement because we do have the type in the expression :)

Anyway, I changed it because it was suggested by @nikola-matic in another PR. I can change it back but I wonder what's his opinion here.

Reply:
Our general policy on ``auto`` was to make references and pointers explicit.

Reply:
Even if you say ``const*`` is redundant here, then so is ``&&```, so then you can just make it a plain ``auto``.

Reply:
So just change it here and anywhere else to ``auto const*``, it's significantly more readable.

Reply:
Change reverted.